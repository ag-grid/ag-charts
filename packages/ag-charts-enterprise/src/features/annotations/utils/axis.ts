import type { _ModuleSupport } from 'ag-charts-community';

export function calculateAxisLabelPadding(axisLayout: _ModuleSupport.AxisLayout) {
    return axisLayout.gridPadding + axisLayout.seriesAreaPadding + axisLayout.tickSize + axisLayout.label.padding;
}
