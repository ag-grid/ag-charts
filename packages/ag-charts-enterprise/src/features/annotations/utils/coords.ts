import { _ModuleSupport, _Scene } from 'ag-charts-community';

import type { AnnotationContext, Point } from '../annotationTypes';
import { convertPoint, invertCoords } from './values';

const { Vec2 } = _ModuleSupport;
const { toRadians } = _Scene;

export function snapPoint(
    offset: _ModuleSupport.Vec2,
    context: AnnotationContext,
    snapping: boolean = false,
    origin?: Point,
    angleStep: number = 1
) {
    if (!snapping) return invertCoords(offset, context);

    const center = origin ? convertPoint(origin, context) : Vec2.origin();
    return invertCoords(snapToAngle(offset, center, angleStep), context);
}

export function snapToAngle(
    { x, y }: _ModuleSupport.Vec2,
    center: _ModuleSupport.Vec2,
    step: number,
    direction: number = 1
) {
    const { x: cx, y: cy } = center;
    const r = Math.sqrt(Math.pow(x - cx, 2) + Math.pow(y - cy, 2));
    const theta = Math.atan2(y - cy, x - cx);

    const stepRadians = toRadians(step);
    const snapTheta = Math.round(theta / stepRadians) * stepRadians;

    return {
        x: cx + r * Math.cos(snapTheta),
        y: cy + r * Math.sin(snapTheta) * direction,
    };
}
