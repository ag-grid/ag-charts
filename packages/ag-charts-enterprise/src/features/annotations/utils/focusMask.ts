import { _ModuleSupport, _Util } from 'ag-charts-community';

const { partialAssign, getWindow, createElement } = _ModuleSupport;
const { copyAttributes } = _Util;

export function createFocusMask(target: HTMLElement): HTMLElement {
    const mask = createElement('div');
    const targetStyle = getWindow().getComputedStyle(target);

    mask.className = 'ag-charts-focus-mask';
    mask.tabIndex = -1;
    mask.style.position = 'absolute';
    mask.style.pointerEvents = 'none';
    partialAssign(['width', 'height'], mask.style, targetStyle);
    copyAttributes(['role', 'aria-labelledby'], mask, target);

    const disappear = () => {
        mask.removeEventListener('keydown', disappear);
        mask.removeEventListener('blur', disappear);
        mask.remove();
    };
    mask.addEventListener('blur', disappear);
    mask.addEventListener('keydown', () => {
        target.focus();
        disappear();
    });

    target.insertAdjacentElement('beforebegin', mask);
    return mask;
}
