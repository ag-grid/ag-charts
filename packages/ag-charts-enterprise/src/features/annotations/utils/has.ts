import { _ModuleSupport } from 'ag-charts-community';

import type {
    AnnotationProperties,
    ChannelPropertiesType,
    LinePropertiesType,
    MeasurerPropertiesType,
    TextualPropertiesType,
} from '../annotationsSuperTypes';
import { CalloutProperties } from '../callout/calloutProperties';
import { CommentProperties } from '../comment/commentProperties';
import { NoteProperties } from '../note/noteProperties';
import { ShapePointProperties } from '../properties/shapePointProperties';
import { isChannelType, isLineType, isMeasurerType, isTextType } from './types';

const { isObject } = _ModuleSupport;

export function hasFontSize(datum?: AnnotationProperties): datum is Exclude<TextualPropertiesType, NoteProperties> {
    return isTextType(datum) && !NoteProperties.is(datum);
}

export function hasLineStyle(
    datum?: AnnotationProperties
): datum is LinePropertiesType | ChannelPropertiesType | MeasurerPropertiesType {
    return isLineType(datum) || isChannelType(datum) || isMeasurerType(datum);
}

export function hasLineColor(datum?: AnnotationProperties) {
    return (
        isLineType(datum) ||
        isChannelType(datum) ||
        isMeasurerType(datum) ||
        CalloutProperties.is(datum) ||
        NoteProperties.is(datum)
    );
}

export function hasIconColor(datum?: AnnotationProperties) {
    return NoteProperties.is(datum);
}

export function hasFillColor(datum?: AnnotationProperties) {
    return (
        isChannelType(datum) ||
        isMeasurerType(datum) ||
        CalloutProperties.is(datum) ||
        CommentProperties.is(datum) ||
        ShapePointProperties.is(datum)
    );
}

export function hasTextColor(datum?: AnnotationProperties) {
    return isTextType(datum) && !NoteProperties.is(datum);
}

export function hasLineText(datum?: AnnotationProperties): datum is LinePropertiesType | ChannelPropertiesType {
    return (isLineType(datum) || isChannelType(datum) || isMeasurerType(datum)) && isObject(datum.text);
}
