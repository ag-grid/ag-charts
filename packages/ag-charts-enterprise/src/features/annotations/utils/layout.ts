import { _Scene } from 'ag-charts-community';

export type PositionedScene = Vec2Scene | Vec4Scene;

type Vec2Scene = _Scene.Node & { x: number; y: number };
type Vec4Scene = _Scene.Node & { x: number; y: number; x1: number; y1: number; x2: number; y2: number };

export function layoutScenesRow(
    scenes: Array<PositionedScene | Array<PositionedScene>>,
    startX: number = 0,
    gap: number = 0
) {
    let x = startX;

    for (const scene of scenes) {
        if (Array.isArray(scene)) {
            for (const scene_ of scene) {
                layoutSetX(scene_, x);
            }
            x += _Scene.Group.computeChildrenBBox(scene).width + gap;
        } else {
            layoutSetX(scene, x);
            x += scene.getBBox().width + gap;
        }
    }
}

export function layoutScenesColumn(
    scenes: Array<PositionedScene | Array<PositionedScene>>,
    startY: number = 0,
    gap: number = 0
) {
    let y = startY;

    for (const scene of scenes) {
        if (Array.isArray(scene)) {
            for (const scene_ of scene) {
                layoutSetY(scene_, y);
            }
            y += _Scene.Group.computeChildrenBBox(scene).height + gap;
        } else {
            layoutSetY(scene, y);
            y += scene.getBBox().height + gap;
        }
    }
}

export function layoutSetX(scene: PositionedScene, x: number) {
    if ('x1' in scene) {
        scene.x2 = x + (scene.x2 - scene.x1);
        scene.x1 = x;
    } else {
        scene.x = x;
    }
}

export function layoutSetY(scene: PositionedScene, y: number) {
    if ('y1' in scene) {
        scene.y2 = y + (scene.y2 - scene.y1);
        scene.y1 = y;
    } else {
        scene.y = y;
    }
}

export function layoutAddX(scene: PositionedScene, x: number) {
    if ('x1' in scene) {
        scene.x1 += x;
        scene.x2 += x;
    } else {
        scene.x += x;
    }
}

export function layoutAddY(scene: PositionedScene, y: number) {
    if ('y1' in scene) {
        scene.y1 += y;
        scene.y2 += y;
    } else {
        scene.y += y;
    }
}
