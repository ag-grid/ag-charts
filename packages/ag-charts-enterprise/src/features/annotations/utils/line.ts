import type { AgAnnotationLineStyleType, PixelSize, _Scene } from 'ag-charts-community';

export function getLineStyle(lineDash?: PixelSize[], lineStyle?: AgAnnotationLineStyleType) {
    return lineDash ? 'dashed' : lineStyle ?? 'solid';
}

export function getComputedLineDash(strokeWidth: number, styleType: AgAnnotationLineStyleType): PixelSize[] {
    switch (styleType) {
        case 'solid':
            return [];
        case 'dashed':
            return [strokeWidth * 4, strokeWidth * 2];
        case 'dotted':
            return [0, strokeWidth * 2];
    }
}

export function getLineDash(
    lineDash?: PixelSize[],
    computedLineDash?: PixelSize[],
    lineStyle?: AgAnnotationLineStyleType,
    strokeWidth?: number
): PixelSize[] | undefined {
    const styleType = getLineStyle(lineDash, lineStyle);
    return computedLineDash ?? lineDash ?? getComputedLineDash(strokeWidth ?? 1, styleType);
}

export function getLineCap(
    lineCap?: _Scene.ShapeLineCap,
    lineDash?: PixelSize[],
    lineStyle?: AgAnnotationLineStyleType
): _Scene.ShapeLineCap | undefined {
    const styleType = getLineStyle(lineDash, lineStyle);
    return lineCap ?? styleType === 'dotted' ? 'round' : undefined;
}
