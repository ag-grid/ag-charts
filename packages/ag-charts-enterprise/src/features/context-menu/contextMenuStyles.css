.ag-chart-context-menu {
    background: rgb(248, 248, 248);
    border: 1px solid #babfc7;
    border-radius: 5px;
    box-shadow: 0 1px 4px 1px rgba(186, 191, 199, 0.4);
    color: rgb(24, 29, 31);
    font:
        13px Verdana,
        sans-serif;
    transition: transform 0.1s ease;
    white-space: nowrap;
    z-index: var(--ag-charts-layer-ui-overlay);
}

.ag-chart-context-menu.ag-charts-dark-context-menu {
    color: white;
    background: #15181c;
}

.ag-chart-context-menu__cover {
    position: fixed;
    left: 0px;
    top: 0px;
}

.ag-chart-context-menu__menu {
    display: flex;
    flex-direction: column;
    padding: 0.5em 0;
}

.ag-chart-context-menu__menu:focus {
    /* Do not show a focus indicator around the context-menu. This makes our custom context-menu
     * behave more like a native context-menu, and makes the behaviour more consistent when using
     * a keyboard or mouse to bring up the context-menu. */
    outline: none;
}

.ag-chart-context-menu__item {
    background: none;
    border: none;
    box-sizing: border-box;
    font: inherit;
    padding: 0.5em 1em;
    text-align: left;
    -webkit-appearance: none;
    -moz-appearance: none;
}

.ag-chart-context-menu__item.ag-charts-dark-context-menu {
    color: white;
}

.ag-chart-context-menu__item:hover {
    background: rgb(33, 150, 243, 0.1);
}

.ag-chart-context-menu__item:hover.ag-charts-dark-context-menu {
    background: rgb(33, 150, 243, 0.1);
}

.ag-chart-context-menu__item:active {
    background: rgb(33, 150, 243, 0.2);
}

.ag-chart-context-menu__item:active.ag-charts-dark-context-menu {
    background: rgb(33, 150, 243, 0.1);
}

.ag-chart-context-menu__item:focus-visible {
    outline: var(--ag-charts-focus-border);
    outline-width: 5px;
    box-shadow: var(--ag-charts-focus-border-shadow);
}

.ag-chart-context-menu__item:focus-visible.ag-charts-dark-context-menu {
    outline: var(--ag-charts-focus-border);
    outline-width: 5px;
    box-shadow: var(--ag-charts-focus-border-shadow);
}

.ag-chart-context-menu__item[aria-disabled='true'] {
    border: none;
    opacity: 0.5;
    text-align: left;
}

.ag-chart-context-menu__item[aria-disabled='true']:hover {
    background: inherit;
    cursor: inherit;
}

.ag-chart-context-menu__divider {
    margin: 5px 0;
    background: #babfc7;
    height: 1px;
}

.ag-chart-context-menu__divider.ag-charts-dark-context-menu {
    background: rgb(33, 150, 243, 0.1);
}
