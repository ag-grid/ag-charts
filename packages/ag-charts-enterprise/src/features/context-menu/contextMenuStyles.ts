export const DEFAULT_CONTEXT_MENU_CLASS = 'ag-chart-context-menu';
export const DEFAULT_CONTEXT_MENU_DARK_CLASS = `ag-charts-dark-context-menu`;

export const defaultContextMenuCss = `
.${DEFAULT_CONTEXT_MENU_CLASS} {
    background: rgb(248, 248, 248);
    border: 1px solid #babfc7;
    border-radius: 5px;
    box-shadow: 0 1px 4px 1px rgba(186, 191, 199, 0.4);
    color: rgb(24, 29, 31);
    font: 13px Verdana, sans-serif;
    transition: transform 0.1s ease;
    white-space: nowrap;
    z-index: 99999;
}

.${DEFAULT_CONTEXT_MENU_CLASS}.${DEFAULT_CONTEXT_MENU_DARK_CLASS} {
    color: white;
    background: #15181c;
}

.${DEFAULT_CONTEXT_MENU_CLASS}__cover {
    position: fixed;
    left: 0px;
    top: 0px;
}

.${DEFAULT_CONTEXT_MENU_CLASS}__menu {
    display: flex;
    flex-direction: column;
    padding: 0.5em 0;
}

.${DEFAULT_CONTEXT_MENU_CLASS}__item {
    background: none;
    border: none;
    box-sizing: border-box;
    font: inherit;
    padding: 0.5em 1em;
    text-align: left;
    -webkit-appearance: none;
    -moz-appearance: none;
}

.${DEFAULT_CONTEXT_MENU_CLASS}__item.${DEFAULT_CONTEXT_MENU_DARK_CLASS} {
    color: white;
}

.${DEFAULT_CONTEXT_MENU_CLASS}__item:hover {
    background: rgb(33, 150, 243, 0.1);
}

.${DEFAULT_CONTEXT_MENU_CLASS}__item:hover.${DEFAULT_CONTEXT_MENU_DARK_CLASS} {
    background: rgb(33, 150, 243, 0.1);
}

.${DEFAULT_CONTEXT_MENU_CLASS}__item:active {
    background: rgb(33, 150, 243, 0.2);
}

.${DEFAULT_CONTEXT_MENU_CLASS}__item:active.${DEFAULT_CONTEXT_MENU_DARK_CLASS} {
    background: rgb(33, 150, 243, 0.1);
}

.${DEFAULT_CONTEXT_MENU_CLASS}__item[disabled] {
    border: none;
    opacity: 0.5;
    text-align: left;
}

.${DEFAULT_CONTEXT_MENU_CLASS}__item[disabled]:hover {
    background: inherit;
    cursor: inherit;
}

.${DEFAULT_CONTEXT_MENU_CLASS}__divider {
    margin: 5px 0;
    background: #babfc7;
    height: 1px;
}

.${DEFAULT_CONTEXT_MENU_CLASS}__divider.${DEFAULT_CONTEXT_MENU_DARK_CLASS} {
    background: rgb(33, 150, 243, 0.1);
}
`;
