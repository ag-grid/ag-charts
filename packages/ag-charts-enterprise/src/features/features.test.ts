import { afterEach, describe, expect, it } from '@jest/globals';

import { type AgChartOptions, AgCharts } from 'ag-charts-community';
import {
    clickAction,
    contextMenuAction,
    dragAction,
    expectWarningsToMatchSnapshots,
    extractImageData,
    hoverAction,
    scrollAction,
    setupMockCanvas,
    setupMockConsole,
    waitForChartStability,
} from 'ag-charts-community-test';

import { prepareEnterpriseTestOptions } from '../test/utils';
import { DEFAULT_CONTEXT_MENU_CLASS } from './context-menu/contextMenuStyles';

describe('Feature Combinations', () => {
    setupMockConsole();

    let chart: any;
    const ctx = setupMockCanvas();

    afterEach(() => {
        if (chart) {
            chart.destroy();
            (chart as unknown) = undefined;
        }
    });

    const compare = async () => {
        await waitForChartStability(chart);

        const imageData = extractImageData(ctx);
        expect(imageData).toMatchImageSnapshot({
            failureThreshold: 0,
            failureThresholdType: 'percent',
        });
    };

    describe('Navigator and Zoom', () => {
        const EXAMPLE_OPTIONS: AgChartOptions = {
            data: [
                { x: 0, y: 0 },
                { x: 1, y: 50 },
                { x: 2, y: 25 },
                { x: 3, y: 75 },
                { x: 4, y: 50 },
                { x: 5, y: 25 },
                { x: 6, y: 50 },
                { x: 7, y: 75 },
            ],
            series: [{ type: 'line', xKey: 'x', yKey: 'y' }],
            navigator: {
                enabled: true,
            },
            zoom: {
                enabled: true,
                axes: 'xy',
                scrollingStep: 0.5,
                minVisibleItemsX: 1,
                minVisibleItemsY: 1,
            },
        };

        let cx: number = 0;
        let cy: number = 0;

        async function prepareChart(
            navigator?: AgChartOptions['navigator'],
            zoom?: AgChartOptions['zoom'],
            baseOptions = EXAMPLE_OPTIONS
        ) {
            const options: AgChartOptions = {
                ...baseOptions,
                navigator: {
                    ...baseOptions.navigator,
                    ...(navigator ?? {}),
                },
                zoom: {
                    ...baseOptions.zoom,
                    ...(zoom ?? {}),
                },
            };
            prepareEnterpriseTestOptions(options);
            cx = options.width! / 2;
            cy = options.height! / 2;

            chart = AgCharts.create(options);

            // Click once in the chart to ensure the chart is active / mouse is over it to ensure the first scroll
            // wheel event is triggered.
            await waitForChartStability(chart);
            await clickAction(cx, cy)(chart);
        }

        it('should zoom then navigate', async () => {
            await prepareChart();

            // Zoom
            await scrollAction(cx, cy, -1)(chart);
            await compare();

            // Move navigator handle
            let from = { x: 417, y: 565 };
            let to = { x: from.x + 100, y: from.y };

            await dragAction(from, to)(chart);

            await compare();

            // Drag navigator slider
            from = { x: to.x + 20, y: to.y };
            to = { x: to.x - 100, y: to.y };

            await hoverAction(from.x, from.y)(chart);
            await dragAction(from, to)(chart);

            await compare();
        });

        it('should init with navigator min/max', async () => {
            await prepareChart({ min: 0.1, max: 0.3 });
            await compare();
        });

        it('should init with zoom ratio', async () => {
            await prepareChart(undefined, { ratioX: { start: 0.7, end: 0.9 } });
            await compare();
        });

        it('should prioritise zoom ratio over navigator min/max', async () => {
            await prepareChart({ min: 0.1, max: 0.3 }, { ratioX: { start: 0.7, end: 0.9 } });
            expectWarningsToMatchSnapshots();
            await compare();
        });

        it('should prioritise zoom range over navigator min/max', async () => {
            await prepareChart({ min: 0.1, max: 0.3 }, { rangeX: { start: 3, end: 6 } });
            await compare();
        });
    });

    describe('Context Menu and Zoom', () => {
        const EXAMPLE_OPTIONS: AgChartOptions = {
            data: [
                { x: 0, y: 0 },
                { x: 1, y: 50 },
                { x: 2, y: 25 },
                { x: 3, y: 75 },
                { x: 4, y: 50 },
                { x: 5, y: 25 },
                { x: 6, y: 50 },
                { x: 7, y: 75 },
            ],
            series: [{ type: 'line', xKey: 'x', yKey: 'y' }],
            contextMenu: {
                enabled: true,
            },
            zoom: {
                enabled: true,
                axes: 'xy',
                scrollingStep: 0.5,
                minVisibleItemsX: 1,
                minVisibleItemsY: 1,
            },
        };

        let cx: number = 0;
        let cy: number = 0;

        async function prepareChart(
            contextMenu?: AgChartOptions['contextMenu'],
            zoom?: AgChartOptions['zoom'],
            baseOptions = EXAMPLE_OPTIONS
        ) {
            const options: AgChartOptions = {
                ...baseOptions,
                contextMenu: {
                    ...baseOptions.contextMenu,
                    ...(contextMenu ?? {}),
                },
                zoom: {
                    ...baseOptions.zoom,
                    ...(zoom ?? {}),
                },
            };
            prepareEnterpriseTestOptions(options);
            cx = options.width! / 2;
            cy = options.height! / 2;

            chart = AgCharts.create(options);

            await waitForChartStability(chart);
            await clickAction(cx, cy)(chart);
        }

        const compareContextMenu = async () => {
            await waitForChartStability(chart);
            expect(document.body.getElementsByClassName(DEFAULT_CONTEXT_MENU_CLASS)).toMatchSnapshot();
        };

        it('when fully zoomed out it should only enable the zoom option', async () => {
            await prepareChart();
            await contextMenuAction(cx, cy)(chart);
            await compareContextMenu();
        });

        it('when zoomed in it should enable both the zoom and pan options', async () => {
            await prepareChart();
            await scrollAction(cx, cy, -1)(chart);
            await contextMenuAction(cx, cy)(chart);
            await compareContextMenu();
        });

        it('when fully zoomed in it should only enable the pan option', async () => {
            await prepareChart();
            await scrollAction(cx, cy, -2)(chart);
            await contextMenuAction(cx, cy)(chart);
            await compareContextMenu();
        });
    });
});
