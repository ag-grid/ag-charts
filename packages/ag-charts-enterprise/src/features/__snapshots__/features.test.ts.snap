// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature Combinations Context Menu and Zoom when fully zoomed in it should only enable the pan option 1`] = `
HTMLCollection [
  <div
    class="ag-chart-context-menu"
    style="display: block; left: 0px; top: calc(0px - 0.5em);"
  >
    <div
      class="ag-chart-context-menu__menu"
    >
      <button
        class="ag-chart-context-menu__item"
        tabindex="-1"
      >
        Download
      </button>
      <button
        class="ag-chart-context-menu__item"
        tabindex="-1"
      >
        Zoom to here
      </button>
      <button
        class="ag-chart-context-menu__item"
        tabindex="-1"
      >
        Pan to here
      </button>
    </div>
  </div>,
]
`;

exports[`Feature Combinations Context Menu and Zoom when fully zoomed out it should only enable the zoom option 1`] = `
HTMLCollection [
  <div
    class="ag-chart-context-menu"
    style="display: block; left: 0px; top: calc(0px - 0.5em);"
  >
    <div
      class="ag-chart-context-menu__menu"
    >
      <button
        class="ag-chart-context-menu__item"
        tabindex="-1"
      >
        Download
      </button>
      <button
        class="ag-chart-context-menu__item"
        tabindex="-1"
      >
        Zoom to here
      </button>
      <button
        class="ag-chart-context-menu__item"
        tabindex="-1"
      >
        Pan to here
      </button>
    </div>
  </div>,
]
`;

exports[`Feature Combinations Context Menu and Zoom when zoomed in it should enable both the zoom and pan options 1`] = `
HTMLCollection [
  <div
    class="ag-chart-context-menu"
    style="display: block; left: 0px; top: calc(0px - 0.5em);"
  >
    <div
      class="ag-chart-context-menu__menu"
    >
      <button
        class="ag-chart-context-menu__item"
        tabindex="-1"
      >
        Download
      </button>
      <button
        class="ag-chart-context-menu__item"
        tabindex="-1"
      >
        Zoom to here
      </button>
      <button
        class="ag-chart-context-menu__item"
        tabindex="-1"
      >
        Pan to here
      </button>
    </div>
  </div>,
]
`;
